### ISSUE TITLE
Aggressive Credential Caching

### ISSUE THREAD

**Issue by SelvamKoch** [aws/aws-toolkit-vscode#2029]
**Describe the bug**

Error suggesting invalid profile on machine with existing credentials in env:variables

**To Reproduce**

After installing and ctrl+shift+p -> connect to AWS on a machine that already has a credentials details in env:variables with MFA enabled. Wait for the session token to expire. Create new credentials and store them in env:variables. Selecting either of env:variables and any other profiles returns an error "Credentials profile is invalid"

**Expected behavior**

AWS connects with env:varaiables when it is updated and valid, since Aggressive caching is fixed in version 1.6.0.

**Screenshots**

![image](https://user-images.githubusercontent.com/90760442/133407804-730c6e7a-0388-4612-aadf-c8a111e27f2e.png)

**Desktop (please complete the following information):**
-   OS: Windows 10
-   Visual Studio Code Version: 1.60.1
-   AWS Toolkit Version: 1.29.0

**Comment by SelvamKoch** [comment]
I understand this is duplicatation of [#705](https://github.com/aws/aws-toolkit-vscode/issues/705) issue. But I'm facing the same issue even in the latest version as well. Looking forward for your help!

**Comment by JadenSimon** [comment]
[This](https://github.com/aws/aws-sdk-js-v3/blob/5e0a46a9f4a35cdb200f7eccef09fb4c6ad76e9c/packages/credential-provider-env/src/index.ts#L14-L30) is what the Toolkit currently uses to get credentials from the environment. Credentials from the environment are 'static' (at least from the Toolkit's perspective) after VS Code starts. Looks like we'll need to spawn a new process to 'reload' the environment variables when applicable.

**Comment by justinmk3** [comment]
> Create new credentials and store them in env:variables. Selecting either of env:variables and any other profiles returns an error "Credentials profile is invalid"

What is updating the environment variables? Processes cannot see env var changes (they are "static" as mentioned above).



> So we'll need to spawn a new process to 'reload' the environment variables when applicable.

What process tree would that be under? I can only imagine it working if we spawn a new shell which happens to get env vars from a ~/.bashrc or other shell init file.

**Comment by JadenSimon** [comment]
> What process tree would that be under? I can only imagine it working if we spawn a new shell which happens to get env vars from a ~/.bashrc or other shell init file.

Yeah I was assuming that the credentials were being loaded in from a shell init script.

**Comment by SelvamKoch** [comment]
> The issue is less so caching, and more with just how environment variables work, though it does look very much like caching from the outside. [This](https://github.com/aws/aws-sdk-js-v3/blob/5e0a46a9f4a35cdb200f7eccef09fb4c6ad76e9c/packages/credential-provider-env/src/index.ts#L14-L30) is what the Toolkit currently uses to get credentials from the environment. Credentials from the environment are 'static' (at least from the Toolkit's perspective) after VS Code starts. So we'll need to spawn a new process to 'reload' the environment variables when applicable.

Thanks for the information. Clear on this now. 
This is not only with env:variables, also with credential file profile as well. In our process, we are updating env variables through shell script once it expired. And then we manually copy those variables in credential file as well. From Ctrl+Shift+P, we  select that profile. Expected behavior is to pick the values from credential and connect to aws. But still it is picking from env. I confirmed by 'aws configure list' command.

My question is, if credential is present in both env and credential file then toolkit will pick from env even we selected profile?

![image](https://user-images.githubusercontent.com/90760442/133501531-43e1a342-f06e-461b-a18e-1b0a548e78f2.png)

**Comment by JadenSimon** [comment]
> My question is, if credential is present in both env and credential file then toolkit will pick from env even we selected profile?

Ah I see. So you have a profile in a credentials file as well?
The toolkit _should_ load the static credentials from the file and ignore the environment variables. Also, the AWS CLI has their own credentials code so I wouldn't worry too much about what it's telling you; the toolkit will favor whatever profile you select.

Are you putting your credentials in `~/.aws/credentials` (or whatever file you have configured) like this?:
```
[default]
aws_access_key_id=[ID]
aws_secret_access_key=[SECRET]
```
The toolkit should then have a `profile:default` option available to select. Choose this instead of `env:variables`.

**Comment by SelvamKoch** [comment]
> The toolkit _should_ load the static credentials from the file and ignore the environment variables. Also, the AWS CLI has their own credentials code so I wouldn't worry too much about what it's telling you; the toolkit will favor whatever profile you select.

This is my understand as well. I have same credentials in both env and [default] profile and selected [default] profile using Ctrl+Shift+P. Toolkit is picking values from env:variables even though we selected [default]. Once we delete the variables from env, it is picking from shared file. 

As of now, I managed to remove env variables and making it work for time being.  Looking forward for your inputs here.

Thanks for the support!

**Comment by JadenSimon** [comment]
> Toolkit is picking values from env:variables even though we selected [default]. Once we delete the variables from env, it is picking from shared file.

I've tried to reproduce this behavior but haven't had any success so far. Does your script write any other environment variables besides `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY`/`AWS_DEFAULT_REGION`? I'm thinking there could be something being set that is causing the JS SDK to ignore what we're telling it to do.

**Comment by SelvamKoch** [comment]
> I've tried to reproduce this behavior but haven't had any success so far, though I'm using just plain old static credentials. Does your script write any other environment variables besides `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY`/`AWS_DEFAULT_REGION`? I'm thinking there could be something being set that is causing the JS SDK to ignore what we're telling it to do.

We are storing aws_access_key_id, aws_secret_access_key, aws_session_token, aws_token_expiration, aws_account_id, aws_account_alias & aws_role in our variables. Not sure if any of this causing the issue.

**Comment by JadenSimon** [comment]
> We are storing aws_access_key_id, aws_secret_access_key, aws_session_token, aws_token_expiration, aws_account_id, aws_account_alias & aws_role in our variables. Not sure if any of this causing the issue.

Okay good to know. Hopefully I can narrow down on which is causing problems.

**Comment by JadenSimon** [comment]
Still have been unable to reproduce the issue, I'm probably missing something.
Here's the steps I followed:
1. Use `aws sts assume-role` to generate temporary credentials with a duration of 900 seconds
2. Export the following variables:
  * `AWS_ACCESS_KEY_ID`
  * `AWS_SECRET_ACCESS_KEY`
  * `AWS_SESSION_TOKEN`
  * `AWS_CREDENTIAL_EXPIRATION`
  * `AWS_ROLE_ARN`
  * `AWS_ROLE_SESSION_NAME`
3. Launch VS Code and use `env:variables` profile
4. Wait 15 minutes for the credentials to expire, then generate a new set
5. Add new credentials to a named profile, then login with the new profile

I've tried a few variations of the steps too and they all work for me. I see that you're using MFA, are you adding MFA-related things to your credential profile or just the plain creds? In particular, `source_profile` may be causing issues, especially if you're using things like `mfa_serial`/`external_id`/`role_arn` that change everytime you change credentials.

**Comment by alknak** [comment]
Agree with @JadenSimon . If you use aws sts to switch roles in on the command line, lacking refresh introduces a pain point. Working around it by closing VSCode and opening it again every time the credentials expire is less fun than it appears.
